core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java

```diff
@@ -528,22 +533,38 @@ private TaskInfo doUpdateTask(
             }
         }
 
-        fragment.map(PlanFragment::getActiveCatalogs)
-                .ifPresent(activeCatalogs -> {
-                    Set<CatalogHandle> catalogHandles = activeCatalogs.stream()
-                            .map(CatalogProperties::getCatalogHandle)
-                            .collect(toImmutableSet());
-                    if (sqlTask.setCatalogs(catalogHandles)) {
-                        ReentrantReadWriteLock.ReadLock catalogInitLock = catalogsLock.readLock();
-                        catalogInitLock.lock();
-                        try {
-                            connectorServicesProvider.ensureCatalogsLoaded(session, activeCatalogs);
-                        }
-                        finally {
-                            catalogInitLock.unlock();
-                        }
+        Optional<List<CatalogProperties>> activeCatalogsOpt = fragment.map(PlanFragment::getActiveCatalogs);
+        if (activeCatalogsOpt.isPresent()) {
+            List<CatalogProperties> activeCatalogs = activeCatalogsOpt.get();
+            Set<CatalogHandle> catalogHandles = activeCatalogs.stream()
+                    .map(CatalogProperties::getCatalogHandle)
+                    .collect(toImmutableSet());
+            if (sqlTask.setCatalogs(catalogHandles)) {
+                ListenableFuture<Void> catalogLoading = Futures.submit(() -> {
+                    ReentrantReadWriteLock.ReadLock catalogInitLock = catalogsLock.readLock();
+                    catalogInitLock.lock();
+                    try {
+                        connectorServicesProvider.ensureCatalogsLoaded(session, activeCatalogs);
                     }
-                });
+                    finally {
+                        catalogInitLock.unlock();
+                    }
+                }, catalogLoadingExecutor);
+                try {
+                    catalogLoading.get(5, SECONDS);
+                    sqlTask.setCatalogsLoaded(immediateVoidFuture());
+                }
```

Comment:
why special casing here with active wait. Why not always use the listener of future.

Reply:
To avoid adding a round trip to most task creates mostly.

Reply:
Most cases catalogs will be "loaded" in well under 5 seconds, because they are already loaded.

Reply:
I am not a big fan of blocking http server thread for 5 seconds. But maybe this is not a big deal. I will let others chime in.

cc: @dain, @sopel39, @findepi